### PR TITLE
Raise appropriate error message when append is not possible

### LINKED ISSUE #1318: Enhance the error message when requesting file format `.db` with sized rotating logger
### Describe the bug

The `can.logger` does not accept `.db` as a format to place the log in, but the API documentation states that `.db` is an acceptable format:
```
    The format is determined from the file suffix which can be one of:
      * .asc: :class:`can.ASCWriter`
      * .blf :class:`can.BLFWriter`
      * .csv: :class:`can.CSVWriter`
      * .db: :class:`can.SqliteWriter`
      * .log :class:`can.CanutilsLogWriter`
      * .txt :class:`can.Printer`
```

### To Reproduce

Enable virtual can: `sudo ip link set vcan0 up type can`
Spam the virtual bus with random can messages 1 ms apart: `cangen vcan0 -e -I 18EAFF26 -L i -v -v -v -g 1`
Log the virtual bus: `python -m can.logger -c vcan0 -i socketcan -b 250000 -f file.db -s 50`

### Expected behavior

The `can.logger` script should successfully log messages. 

### Additional context

OS and version: Ubuntu 20.04.4 LTS
Python version: Python 3.7.17
python-can version: 4.0.0

<details><summary>Traceback and logs</summary>

```python
Traceback (most recent call last):
  File "/home/agco/Documents/python-can/scripts/can_logger.py", line 11, in <module>
    main()
  File "/home/agco/Documents/python-can/can/logger.py", line 206, in main
    base_filename=results.log_file, max_bytes=results.file_size
  File "/home/agco/Documents/python-can/can/io/logger.py", line 322, in __init__
    self._writer = self._get_new_writer(self.base_filename)
  File "/home/agco/Documents/python-can/can/io/logger.py", line 230, in _get_new_writer
    "The Logger corresponding to the arguments is not a FileIOMessageWriter or "
Exception: The Logger corresponding to the arguments is not a FileIOMessageWriter or can.Printer
```
</details>

### PR DESCRIPTION
**Pull request by j-c-cook**
- Handle "append" by adding **options to SqlWriter 
- Check logger instance for SqliteWriter in rotating logger setup and raise appropriate error.

closes #1318

### CODE REVIEW

**Comment by j-c-cook** [comment]
@zariiii9003 ready for review.

File: can/io/sqlite.py [hunk 1]
```diff
@@ -128,7 +128,9 @@ class SqliteWriter(MessageWriter, BufferedReader):
     MAX_BUFFER_SIZE_BEFORE_WRITES = 500
     """Maximum number of messages to buffer before writing to the database"""
 
-    def __init__(self, file: StringPathLike, table_name: str = "messages") -> None:
+    def __init__(
+        self, file: StringPathLike, table_name: str = "messages", **options: Any
```
Comment (zariiii9003) [inline can/io/sqlite.py:132]:
```suggestion
        self, file: StringPathLike, table_name: str = "messages", **kwargs: Any
```

This is usually called `kwargs` in python projects

File: can/io/logger.py [hunk 2]
```diff
@@ -225,6 +226,11 @@ def _get_new_writer(self, filename: StringPathLike) -> FileIOMessageWriter:
...
         elif isinstance(logger, Printer) and logger.file is not None:
             return cast(FileIOMessageWriter, logger)
+        elif isinstance(logger, can.io.SqliteWriter):
+            raise Exception(
+                "The SqliteWriter is not available in conjunction"
+                "with the sized rotating logger."
+            )
         else:
             raise Exception(
                 "The Logger corresponding to the arguments is not a FileIOMessageWriter or "
```
Comment (zariiii9003) [inline can/io/logger.py:236]: You could improve the existing error message without the `isinstance(logger, can.io.SqliteWriter)` check. You can access the logger class name with `logger.__class__.__name__`, for example inside a f-string.
This way you do not need special cases for any new unsupported loggers

**Review by zariiii9003 — CHANGES_REQUESTED** [review]

File: can/io/sqlite.py (hunk 1 above)
Comment (j-c-cook) [reply can/io/sqlite.py:132]: > Also: Shouldn't there be an exception when the user demands "append" but the logger does not support it?

I think this is brilliant.

**Comment by j-c-cook** [comment]
@zariiii9003 Thank you for that feedback. I added in the "append" error handling to `ASCWriter` and `SQLWriter` and adjusted the exception based on your [comment](https://github.com/hardbyte/python-can/pull/1361#discussion_r939557531).

File: can/io/asc.py [hunk 3]
```diff
@@ -360,6 +361,13 @@ def __init__(
         :param channel: a default channel to use when the message does not
                         have a channel set
         """
+        if "append" in kwargs:
+            if kwargs["append"]:
+                raise TypeError(
```
Comment (j-c-cook) [inline can/io/asc.py:366]: I'm not sure if this is the best kind of error to use. Maybe a `ValueError`?

Comment (zariiii9003) [reply can/io/asc.py:366]: Yes, I'd say ValueError, too

File: can/io/logger.py [hunk 4]
```diff
@@ -227,8 +228,8 @@ def _get_new_writer(self, filename: StringPathLike) -> FileIOMessageWriter:
             return cast(FileIOMessageWriter, logger)
         else:
             raise Exception(
-                "The Logger corresponding to the arguments is not a FileIOMessageWriter or "
-                "can.Printer"
+                f"The Logger, {logger.__class__.__name__}, corresponding to "
```
Comment (zariiii9003) [inline can/io/logger.py:231]: The previous error message wasn't great. We should use simple language.

```python3
raise Exception(
    f"The log format \"{''.join(pathlib.Path(filename).suffixes[-2:])}\" "
    f"is not supported by {self.__class__.__name__}"
)
```
This results in `Exception: The log format ".db" is not supported by SizedRotatingLogger`.
Works for all subclasses, too. We could also point the user to the documentation but that might be a bit redundant.

File: can/io/asc.py [hunk 5]
```diff
@@ -360,6 +361,13 @@ def __init__(
         :param channel: a default channel to use when the message does not
                         have a channel set
         """
+        if "append" in kwargs:
+            if kwargs["append"]:
```
Comment (zariiii9003) [inline can/io/asc.py:365]: you could remove this line if you use `if kwargs.get("append", False):`

File: can/io/asc.py [hunk 6]
```diff
@@ -361,13 +358,11 @@ def __init__(
...
-                raise TypeError(
-                    f"{self.__class__.__name__} is currently not "
-                    f"equipped to append messages to an existing "
-                    f"file."
-                )
+        if kwargs.get("append", False):
+            raise TypeError(
+                f"{self.__class__.__name__} is currently not equipped to "
+                f"append messages to an existing file."
+            )
```
Comment (j-c-cook) [inline can/io/asc.py:368]: Took your advice here.

File: can/io/logger.py [hunk 7]
```diff
@@ -228,7 +228,7 @@ def _get_new_writer(self, filename: StringPathLike) -> FileIOMessageWriter:
             return cast(FileIOMessageWriter, logger)
         else:
             raise Exception(
-                f"The log format \".{filename.split('.')[-1]}\" is not "
+                f"The log format \"{os.path.splitext(filename)[-1]}\" is not "
```
Comment (j-c-cook) [inline can/io/logger.py:231]: I used your recommended error message, but slightly adjusted how the extension is found.

Comment (zariiii9003) [reply can/io/logger.py]: This does not work for multiple suffixes like `file.asc.gz`

Comment (j-c-cook) [reply can/io/logger.py]: I fixed it with your version.

**Comment by j-c-cook** [comment]
@zariiii9003 I think this might be ready to go in. Could you give it another look?

**Review by zariiii9003 — APPROVED** [review]
Thank you, looks good

**Pull request merged**